src/boot/efi/linux_x86.c

```diff
@@ -184,15 +184,30 @@ EFI_STATUS linux_exec_efi_handover(
                 boot_params->hdr.setup_sects = 4;
 
         _cleanup_pages_ Pages cmdline_pages = {};
-        if (cmdline) {
+        if (load_options) {
+                size_t cmdline_len = MIN(load_options_size / sizeof(char16_t), image_params->hdr.cmdline_size);
+
                 cmdline_pages = xmalloc_pages(
                                 can_4g ? AllocateAnyPages : AllocateMaxAddress,
                                 EfiLoaderData,
                                 EFI_SIZE_TO_PAGES(cmdline_len + 1),
                                 CMDLINE_PTR_MAX);
+                char *cmdline = PHYSICAL_ADDRESS_TO_POINTER(cmdline_pages.addr);
+
+                for (size_t i = 0; i < cmdline_len; i++) {
+                        char16_t c = ((char16_t *) load_options)[i];
+
+                        /* Kernel EFI stub ends processing LoadOptions on these. */
+                        if (IN_SET(c,'\0', '\n')) {
+                                cmdline_len = i;
+                                break;
```

Comment:
so we hade reports previously, where people used kernel command lines with embedded newlines and expected them to work (because people did `echo foo > /etc/kernel/cmdline`). we probably should convert them to spaces. which the status quo ante did.

Reply:
i.e #7079

Reply:
Btw, that link is wrong.